src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -701,7 +701,17 @@ internal void ProcessMessage(IBackendMessage msg)
 
         void ProcessDataRowMessage(DataRowMessage msg)
         {
-            Connector.State = ConnectorState.Fetching;
+            // Protection from concurrent break
+            // See #1502
+            lock (Connector)
+            {
+                if (Connector.State == ConnectorState.Broken)
+                {
+                    throw Connector.BreakReason!;
```

Comment:
Rethrowing an exception overrides the original stack trace. Use `EceptionDispatchInfo` to hold an exception and to rethrow it.